### PR TITLE
gh-121797: Add class method Fraction.from_number()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is an alternative constructor which only accepts a single numeric argument. Unlike to Fraction.from_float() and Fraction.from_decimal() it accepts any real numbers supported by the standard constructor (int, float, Decimal, Rational numbers).
Unlike to the standard constructor, it does not accept strings.

* Issue: gh-121797

----
📚 Documentation preview 📚: https://cpython-previews--121800.org.readthedocs.build/

### CODE REVIEW

**Review by mdickinson — APPROVED** [review]
+1 for the general idea. I haven't reviewed line-by-line.

File: Lib/fractions.py [hunk 1]
```diff
@@ -305,6 +306,28 @@ def __new__(cls, numerator=0, denominator=None):
         self._denominator = denominator
         return self
 
+    @classmethod
+    def from_number(cls, number):
```
Comment (skirpichev) [inline Lib/fractions.py:310]: Hmm, you can reuse this in the ~~Rational~~Fraction constructor, isn't?  First three if/elif's share exactly same logic.

**Review by skirpichev — APPROVED** [review]
LGTM with one nitpick (feel free to ignore)

File: Lib/fractions.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/fractions.py:310]: Do you mean the main Fraction constructor?

There is a different exception if the argument type is not supported. There is also a matter of performance.

Comment (skirpichev) [reply Lib/fractions.py:310]: I meant some helper:
```py
@classmethod
def _from_number(cls, number):
    if type(number) is int:
        return cls._from_coprime_ints(number, 1)
    elif isinstance(number, numbers.Rational):
        return cls._from_coprime_ints(number.numerator, number.denominator)
    elif (isinstance(number, float) or
          (not isinstance(number, type) and
           hasattr(number, 'as_integer_ratio'))):
        return cls._from_coprime_ints(*number.as_integer_ratio())
```

Then you can reuse one in the constructor and in the ``from_number()``, e.g.:
```
        if denominator is None:
            self = cls._from_number(numerator)
            if self is not None:
                return self

            elif isinstance(numerator, str):            
                ...
```

> There is also a matter of performance.

I don't expect too much from extra class method call.

Comment (serhiy-storchaka) [reply Lib/fractions.py:310]: Extra function call and extra check.

This will not save much code. I'll leave this on future if we touch this code again.

Comment (skirpichev) [reply Lib/fractions.py:310]: Well, it's up to you.  PR has Mark approval, so it's ready to go.

**Pull request merged**